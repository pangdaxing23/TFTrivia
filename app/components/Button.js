import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableHighlight,
  TouchableNativeFeedback,
  Platform,
} from 'react-native';

type Props = {
  text?: string,
  onPress: Function,
  disabled?: boolean
};

const Button = ({ text, onPress, disabled }: Props) => {
  if (Platform.OS === 'ios') {
    return (
      <TouchableHighlight
        onPress={disabled ? () => {} : onPress}
        style={[styles.button, disabled ? styles.disabled : styles.enabled]}
        underlayColor={underlayColor}
      >
        <Text style={styles.text}>{text}</Text>
      </TouchableHighlight>
    );
  } else {
    return (
      <TouchableNativeFeedback
        onPress={disabled ? () => {} : onPress}
        background={TouchableNativeFeedback.Ripple(underlayColor)}
      >
        <View
          style={[
            styles.button,
            styles.android,
            disabled ? styles.disabled : styles.enabled,
          ]}
        >
          <Text style={styles.text}>{text}</Text>
        </View>
      </TouchableNativeFeedback>
    );
  }
};

const underlayColor = '#418CED';
const styles = StyleSheet.create({
  button: {
    paddingHorizontal: 40,
    paddingVertical: 15,
    marginHorizontal: 10,
    borderRadius: 10,
    minWidth: 145,
  },
  android: {
    borderRadius: 4,
    elevation: 2,
  },
  enabled: {
    backgroundColor: '#6BC1E9',
  },
  disabled: {
    backgroundColor: '#444444',
  },
  text: {
    textAlign: 'center',
    fontSize: 22,
    color: '#FFF',
  },
});

export default Button;
