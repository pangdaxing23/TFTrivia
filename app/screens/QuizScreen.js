import React, { Component } from 'react';
import { Text, View } from 'react-native';
import Container from '../components/Container';
import Card from '../components/Card';
import Tally from '../components/Tally';
import Answers from '../components/Answers';
import styles from '../styles';

type Props = {
  navigation: {
    state: {
      params: {
        category: string,
        question: string,
        count: number,
        total: number,
        onPressFalse?: Function,
        onPressTrue?: Function
      }
    }
  }
};

export default class QuizScreen extends Component<Props> {
  render () {
    const {
      category,
      question,
      count,
      total,
      onPressTrue,
      onPressFalse,
    } = this.props.navigation.state.params;
    return (
      <Container>
        <Text style={styles.head}>{category}</Text>
        <View>
          <Card text={question} />
        </View>
        <Tally count={count} total={total} />
        <Answers onPressTrue={onPressTrue} onPressFalse={onPressFalse} />
      </Container>
    );
  }
}
