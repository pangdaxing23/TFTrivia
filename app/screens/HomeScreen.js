import React, { Component } from 'react';
import { Text, View } from 'react-native';
import Container from '../components/Container';
import Button from '../components/Button';
import { NavigationActions } from 'react-navigation';
import styles from '../styles';
import { fetchData } from '../config/api';
import he from 'he';

type Props = {
  navigation: Object
};

export default class HomeScreen extends Component<Props> {
  constructor (props) {
    super();
    this.state = {
      data: {},
      loaded: false,
      index: 0,
      correct: 0,
    };
  }

  componentDidMount = async () => {
    let data = await fetchData();
    data = data.results;
    this.setState(
      state => data,
      () => {
        this.setState({ loaded: true });
      }
    );
    this.setState({ data });
  };

  nextScreen = () => {
    const { data, index } = this.state;
    const total = data.length;

    if (index === data.length) {
      this.props.navigation.navigate('ResultsScreen', {
        correct: this.state.correct,
        total,
        onPlayAgain: this.onPlayAgain,
      });
    } else {
      let { category, question } = data[index];
      question = he.decode(question);
      const count = index + 1;

      this.props.navigation.navigate('QuizScreen', {
        category,
        question,
        count,
        total,
        onPressTrue: this.onPressTrue,
        onPressFalse: this.onPressFalse,
      });
    }
  };

  onPressTrue = () => {
    let currentQuestion = this.state.data[this.state.index];
    if (currentQuestion.correct_answer === 'True') {
      this.setState({
        correct: this.state.correct + 1,
      });
    }
    // must be callback version of setState because we read
    // this.state.index in nextScreen
    this.setState(this.incrementIndex, this.nextScreen);
  };

  onPressFalse = () => {
    let currentQuestion = this.state.data[this.state.index];
    if (currentQuestion.correct_answer === 'False') {
      this.setState({
        correct: this.state.correct + 1,
      });
    }
    this.setState(this.incrementIndex, this.nextScreen);
  };

  onPlayAgain = () => {
    // reset the navigation stack
    const resetAction = NavigationActions.reset({
      index: 0,
      actions: [NavigationActions.navigate({ routeName: 'HomeScreen' })],
    });
    this.props.navigation.dispatch(resetAction);
  };

  incrementIndex (state) {
    return { index: state.index + 1 };
  }

  render () {
    return (
      <Container>
        <Text style={styles.head}>{texts.welcome}</Text>
        <View>
          <Text style={styles.info}>{texts.info}</Text>
          <Text style={styles.info}>{texts.question}</Text>
        </View>
        <Button
          text={texts.begin}
          onPress={this.nextScreen}
          disabled={!this.state.loaded}
        />
      </Container>
    );
  }
}

const texts = {
  welcome: 'Welcome to the Trivia Challenge!',
  info: 'You will be presented with 10 True or False questions.',
  question: 'Can you score 100%?',
  begin: 'BEGIN',
};
