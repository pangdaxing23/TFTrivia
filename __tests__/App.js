import 'react-native';
import React from 'react';
import HomeScreen from '../app/screens/HomeScreen';
import QuizScreen from '../app/screens/QuizScreen';
import Button from '../app/components/Button';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders HomeScreen', () => {
  const homeScreen = renderer.create(<HomeScreen />);

  expect(homeScreen).toMatchSnapshot();
});

it('renders QuizScreen', () => {
  // TODO: this https://github.com/vonovak/react-navigation-props-mapper may be
  // simple solution to this
  const quizScreen = renderer.create(
    <QuizScreen
      category='Lemons'
      question='How many lemons are too many?'
      count={3}
      total={10}
    />
  );

  expect(quizScreen).toMatchSnapshot();
});

it('button renders text', () => {
  const button = renderer.create(<Button text='Fangorn' />);

  expect(button).toMatchSnapshot();
});

it('button renders when disabled', () => {
  const button = renderer.create(<Button text='Fangorn' disabled />);

  expect(button).toMatchSnapshot();
});
