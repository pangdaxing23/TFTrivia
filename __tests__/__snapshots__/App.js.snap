// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button renders text 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Array [
        Object {
          "borderRadius": 10,
          "marginHorizontal": 10,
          "minWidth": 145,
          "paddingHorizontal": 40,
          "paddingVertical": 15,
        },
        Object {
          "backgroundColor": "#6BC1E9",
        },
      ],
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#FFF",
        "fontSize": 22,
        "textAlign": "center",
      }
    }
  >
    Fangorn
  </Text>
</View>
`;

exports[`button renders when disabled 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Array [
        Object {
          "borderRadius": 10,
          "marginHorizontal": 10,
          "minWidth": 145,
          "paddingHorizontal": 40,
          "paddingVertical": 15,
        },
        Object {
          "backgroundColor": "#444444",
        },
      ],
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#FFF",
        "fontSize": 22,
        "textAlign": "center",
      }
    }
  >
    Fangorn
  </Text>
</View>
`;

exports[`renders HomeScreen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#A2E4F9",
      "flex": 1,
      "justifyContent": "space-around",
      "paddingHorizontal": 30,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000",
        "fontSize": 24,
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    Welcome to the Trivia Challenge!
  </Text>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontSize": 22,
          "marginBottom": 50,
          "textAlign": "center",
        }
      }
    >
      You will be presented with 10 True or False questions.
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontSize": 22,
          "marginBottom": 50,
          "textAlign": "center",
        }
      }
    >
      Can you score 100%?
    </Text>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "borderRadius": 10,
            "marginHorizontal": 10,
            "minWidth": 145,
            "paddingHorizontal": 40,
            "paddingVertical": 15,
          },
          Object {
            "backgroundColor": "#444444",
          },
        ],
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFF",
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      BEGIN
    </Text>
  </View>
</View>
`;

exports[`renders QuizScreen 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#A2E4F9",
      "flex": 1,
      "justifyContent": "space-around",
      "paddingHorizontal": 60,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 40,
        "textAlign": "center",
      }
    }
  >
    Lemons
  </Text>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "backgroundColor": "#fff",
          "padding": 25,
        }
      }
    >
      How many lemons are too many?
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000",
        "fontSize": 21,
        "marginBottom": 50,
        "textAlign": "center",
      }
    }
  >
    3
    /
    10
  </Text>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Object {
          "backgroundColor": "#6BC1E9",
          "borderRadius": 10,
          "paddingHorizontal": 40,
          "paddingVertical": 15,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#FFF",
            "fontSize": 22,
          }
        }
      >
        BEGIN
      </Text>
    </View>
  </View>
</View>
`;
